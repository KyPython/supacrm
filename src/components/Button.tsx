"use client";
import React from "react";
import Link from "next/link";
import { IconType } from "react-icons";

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  variant?: "primary" | "secondary" | "ghost";
  className?: string;
  leftIcon?: React.ReactNode | IconType;
  rightIcon?: React.ReactNode | IconType;
  ariaLabel?: string;
  href?: string;
};

function renderIcon(icon?: React.ReactNode | IconType) {
  if (!icon) return null;
  if (typeof icon === "function") {
    const C = icon as IconType;
    return (
      <span className="btn-icon" aria-hidden>
        <C />
      </span>
    );
  }
  return (
    <span className="btn-icon" aria-hidden>
      {icon}
    </span>
  );
}

export default function Button({
  children,
  variant = "primary",
  className = "",
  leftIcon,
  rightIcon,
  ariaLabel,
  href,
  ...props
}: ButtonProps) {
  const base =
    "inline-flex items-center justify-center gap-2 px-4 py-2 rounded focus:outline-none focus:ring-2 focus:ring-offset-2";

  const style: React.CSSProperties = {};
  if (variant === "primary") {
    style.background = "var(--brand)";
    style.color = "var(--fg)";
  }
  if (variant === "secondary") {
    style.background = "var(--card)";
    style.color = "var(--fg)";
  }

  const disabledClass = (props as any).disabled
    ? "opacity-60 cursor-not-allowed"
    : "hover:shadow-md";

  const iconLeft = renderIcon(leftIcon);
  const iconRight = renderIcon(rightIcon);

  if (href) {
    const isInternal = href.startsWith("/");
    const sharedProps = {
      "aria-label": ariaLabel,
      className: `${base} ${disabledClass} ${className}`.trim(),
      style,
    } as any;

    if (isInternal) {
      // Use Next.js Link for client-side navigation on internal routes
      const { onClick, target, rel } = props as any;
      return (
        <Link
          href={href}
          {...sharedProps}
          onClick={onClick}
          target={target}
          rel={rel}
        >
          {iconLeft}
          <span>{children}</span>
          {iconRight}
        </Link>
      );
    }

    // external link - keep anchor behavior
    return (
      <a
        aria-label={ariaLabel}
        className={`${base} ${disabledClass} ${className}`.trim()}
        style={style}
        href={href}
        {...(props as any)}
      >
        {iconLeft}
        <span>{children}</span>
        {iconRight}
      </a>
    );
  }

  return (
    <button
      aria-label={ariaLabel}
      className={`${base} ${disabledClass} ${className}`.trim()}
      style={style}
      {...props}
    >
      {iconLeft}
      <span>{children}</span>
      {iconRight}
    </button>
  );
}
