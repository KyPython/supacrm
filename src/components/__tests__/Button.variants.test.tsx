import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Button from "../Button";
import { HiOutlinePlus } from "react-icons/hi";

describe("Button variants", () => {
  test("secondary snapshot and click", async () => {
    const handle = jest.fn();
    render(
      <Button variant="secondary" onClick={handle}>
        Secondary
      </Button>
    );
    const btn = screen.getByRole("button", { name: /secondary/i });
    await userEvent.click(btn);
    expect(handle).toHaveBeenCalled();
  });

  test("ghost anchor renders as link", () => {
    const { container } = render(
      <Button href="/test" variant="ghost">
        Link
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  test("left and right icons render", () => {
    const { container } = render(
      <Button leftIcon={HiOutlinePlus} rightIcon={<span>R</span>}>
        WithIcons
      </Button>
    );
    expect(container).toMatchSnapshot();
  });
});
