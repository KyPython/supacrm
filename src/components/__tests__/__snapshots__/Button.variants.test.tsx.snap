// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button variants ghost anchor renders as link 1`] = `
<div>
  <a
    class="inline-flex items-center justify-center gap-2 px-4 py-2 rounded focus:outline-none focus:ring-2 focus:ring-offset-2 hover:shadow-md"
    href="/test"
  >
    <span>
      Link
    </span>
  </a>
</div>
`;

exports[`Button variants left and right icons render 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 px-4 py-2 rounded focus:outline-none focus:ring-2 focus:ring-offset-2 hover:shadow-md"
  >
    <span
      aria-hidden="true"
      class="btn-icon"
    >
      <svg
        aria-hidden="true"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 4v16m8-8H4"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </span>
    <span>
      WithIcons
    </span>
    <span
      aria-hidden="true"
      class="btn-icon"
    >
      <span>
        R
      </span>
    </span>
  </button>
</div>
`;
